refactor: Introduce the (outlines of) the Firestore Native-based v2 backend

rjhuijsman opened the pull request:
This PR lays the groundwork for the v2 backend, which will be based on Firestore in Native mode.

Reviewer suggestion:
1. Begin with the `README.md` files to get some idea of what the moving pieces are.
1. Look at the current Firestore Security Rules in `app/server/firestore.rules` and its unit tests in `app/server/functions/firestore_rules.spec.ts`.
1. Look at the `app/server/functions/src/index.ts` file containing the new Cloud Functions.
1. Look at the development setup for the Cloud Functions in `app/server/functions/package.json`.
1. Look at the changes to the Terraform config in `app/server/terraform`.
1. Look at how Firebase specifies what it will deploy in `app/server/firebase.json`.
1. Optionally, try deploying a v2 backend for yourself, and let me know how it goes!

## Screenshots
None

## How did you test the change?

- [ ] iOS Simulator
- [ ] iOS Device
- [ ] Android Simulator
- [ ] Android Device
- [X] `curl` to a dev App Engine server
- [X] other, please describe: Unit tests, and manual verification using Google Cloud Console.

## Checklist:

- [X] Followed the [Contributor Guidelines](https://github.com/WorldHealthOrganization/app/blob/master/docs/CONTRIBUTING.md) and verified that all contributions are properly licensed pursuant to the [LICENSE](https://github.com/WorldHealthOrganization/app/blob/master/LICENSE).

matthewblain commented on server/functions/src/index.ts:
```diff
@@ -0,0 +1,126 @@
+  .https.onRequest((request, response) => {
+    // A "fake" of the v1 API: we're just going to return a static file that we
+    // happen to have on disk.
+    // TODO: replace this with actual statistics in Firestore.
```
for sure! given that the current thing puts stats in datastore, do you think it's easier to modify the app engine app to write to datastore-in-firestore-mode or just rewrite? would the rewrite be in typescript? It's pretty straightforward.

rjhuijsman replied:
I suggest we rewrite the existing App Engine code as a [scheduled Cloud Function](https://firebase.google.com/docs/functions/schedule-functions). That way we don't have to introduce AppEngine to the v2 backend at all.

brunobowden replied:
@rjhuijsman - I like the idea of using the Scheduled Cloud Function too

matthewblain commented on server/functions/src/index.ts:
```diff
@@ -0,0 +1,126 @@
+// Implementation of the v1 API"s `getCaseStats` method.
+// TODO: replace with direct Firestore access from the client.
+export const getCaseStats = functions
+  .region("europe-west6")
```
configuration question: what does this do? the code is specifying a datacenter which it runs in somehow?  what happens if we want to run ain instance in us-north-27 or whatever? (note: unlikley to happen but... still seems odd)

rjhuijsman replied:
Yeah, this indeed is "infrastructure configuration as code" - it's analogous to writing a Terraform config telling us where to deploy the code, but does it inline in the source code itself.

If we always want to deploy to multiple regions, we can also pass a list of regions.

If we want to deploy some project IDs to region X and others to region Y, we can compute the value given to "region" inline before exporting the function.

If we hate the inline configuration inside source code, we can drop the Firebase tooling and deploy the Cloud Functions through standard `gcloud` tooling and Terraform.

matthewblain commented on server/functions/src/index.ts:
```diff
@@ -0,0 +1,126 @@
+
+// Implementation of the v1 API"s `putClientSettings` method.
+// TODO: replace with direct Firestore acccess from the client.
+export const putClientSettings = functions
```
ah, so this is a full on drop in server replacement api for now.

matthewblain commented on server/functions/src/index.ts:
```diff
@@ -0,0 +1,126 @@
+    // Get the previous document value (for update or delete).
+    const oldDoc = change.before.data();
+
+    // TODO: replace the following placeholder code with real implementation
```
This actually deserves a redesign since no one is using it. The current scheme is not localized. But also can be done as one of the last steps since no one is using it.

matthewblain replied:
Also: is this best done client side or server side? Classic api hiding question.

rjhuijsman replied:
Ah, good to know! I wasn't aware that the notifications system wasn't being used at all. Let's discuss this as a separate PR.

matthewblain commented on server/public/404.html:
```diff
@@ -0,0 +1,33 @@
+      <h2>404</h2>
+      <h1>Page Not Found</h1>
+      <p>The specified file was not found on this website. Please check the URL for mistakes and try again.</p>
+      <h3>Why am I seeing this?</h3>
```
hah! Should this page be even simpler? Or just delete these two lines.

rjhuijsman replied:
Removed the two lines for now. Happy to have a proposal for anything even better though!

matthewblain commented on server/public/index.html:
```diff
@@ -0,0 +1,89 @@
+<!DOCTYPE html>
```
same thing -delete/ replace with 404 page?
can we put in a '302' page or maybe an html meta refresh redirect for the two 'app store' pages?

rjhuijsman replied:
We can definitely do redirects. I need to read up on how exactly, so I'm going to leave it as a TODO in the `README.md`'s Development Plan section for now if that's OK.

matthewblain commented on server/terraform/README.md:
```diff
@@ -182,11 +183,13 @@ emacs main.tf
 terraform apply
-# Should create all resources without any errors
+# THIS MAY THROW AN ERROR:
+#   Error: googleapi: Error 409: Sink _Default already exists, alreadyExists
```
yeah terraform is strange. i should dig up my notes on this for dev.

how does logging work in firestore anyway? if we remove app engine and cloud storage, what logs do we even see?

rjhuijsman replied:
We'll get logs from the Cloud Functions, those output logs roughly the same way that AppEngine does. Firestore will log deploys (changes to the Rules), but not access requests.

matthewblain commented on server/terraform/modules/myhealth/main.tf:
```diff
@@ -97,11 +98,14 @@ resource "google_firebase_project_location" "fireloc" {
+# App Engine - required to create the Firestore database.
+#   See: https://firebase.google.com/docs/firestore/solutions/automate-database-create#create_a_database_with_terraform
+#   NOTE: contrary to this documentation, we seem to still need a manual step to actually create the Firestore database.
+#         That manual step is documented in `server/terraform/README.md`.
```
can it be done in gcloud?

rjhuijsman replied:
Surprisingly, no! Using `gcloud` calls the same APIs that Terraform does, which makes it even stranger that neither of those work! The only way I've found to do this is actually click around on the webpage.

matthewblain commented on server/terraform/v2-dev1/.terraform.lock.hcl:
```diff
@@ -0,0 +1,92 @@
+# This file is maintained automatically by "terraform init".
```
.gitignore

rjhuijsman replied:
Done.

matthewblain commented on server/terraform/modules/myhealth/main.tf:
```diff
@@ -69,6 +69,7 @@ resource "google_project_service" "service" {
     "compute.googleapis.com",
     "dns.googleapis.com",
     "firebase.googleapis.com",
+    var.backend_version == "v2" ? "firebasehosting.googleapis.com" : null,
```
I wonder if it's clearner to simply fork this to myhelathv2/main.tf. May be much cleaner to manages.

rjhuijsman replied:
Done.

matthewblain reviewed (commented):
Looks like the right direction. My comments are mostly procedural/naive, as I've never used firestore or typescript.

brunobowden commented on server/WHO.postman_collection.json:
```diff
@@ -0,0 +1,174 @@
+{
```
Possible to replace tabs with double spaces? Then it would show better on GitHub during review and editing.

rjhuijsman replied:
Done.

rjhuijsman replied:
Done.

brunobowden replied:
@rjhuijsman - when an issue is "Done", just click on the "Resolve conversation" in GitHub

brunobowden commented on server/fetch-static-serving.sh:
```diff
@@ -0,0 +1,12 @@
+#!/bin/bash
+set -ev
```
```suggestion
set -euv
```

rjhuijsman replied:
Done.

brunobowden commented on server/functions/casestats.json:
```diff
@@ -0,0 +1,2781 @@
+{
+    "jurisdictionStats": [
+        {
```
Prefer 2 spaces for indentation but not a big issue

rjhuijsman replied:
Done

brunobowden commented on server/functions/casestats.json:
```diff
@@ -0,0 +1,2781 @@
+{
```
This file doesn't appear to be used anywhere in this PR

rjhuijsman replied:
It's used to feed the "fake" `getCaseStats` API; see its use in `index.ts`.

We should remove this ASAP though! 🙂

rjhuijsman replied:
I'm going to choose to be lazy here and leave it as-is, with the intention of removing this file again ASAP when we do #1972.

brunobowden commented on server/functions/src/firestore_rules.spec.ts:
```diff
@@ -0,0 +1,41 @@
+import * as firebase from "@firebase/rules-unit-testing";
+import "mocha";
```
Is mocha used?

rjhuijsman replied:
Not here; gracias.

brunobowden commented on server/functions/src/firestore_rules.spec.ts:
```diff
@@ -0,0 +1,41 @@
+import * as firebase from "@firebase/rules-unit-testing";
+import "mocha";
+
+// Connect to the local Firestore emulator. The Firestore tooling sets
```
Add description.... I think this is for testing the firebase rules?

rjhuijsman replied:
Description added.

brunobowden commented on server/functions/src/index.ts:
```diff
@@ -0,0 +1,126 @@
+// Method that runs when client settings have been updated, and will make those
+// updated settings take effect.
+export const clientSettingsUpdated = functions
+  .region("europe-west6")
```
Extract as common variable? Can this be configured at a single place in the terraform code, so that it's simple to change the location of the entire system? Some countries may prefer deploying it within their continent for example.

See single location for the Terraform code:
https://github.com/WorldHealthOrganization/app/blob/master/server/terraform/modules/myhealth/variables.tf#L13

rjhuijsman replied:
We can configure it in a single place in the Firebase code, but it's tricky to share that with Terraform as it currently is. If we care a lot about keeping it in Terraform only, we may find it easier to drop the Firebase tooling for Cloud Functions and use "vanilla" Cloud Functions controlled by a Terraform spec. A bit more work and code, but it would centralize that config value.

brunobowden commented on server/functions/src/index.ts:
```diff
@@ -0,0 +1,126 @@
+// Implementation of the v1 API"s `putClientSettings` method.
+// TODO: replace with direct Firestore acccess from the client.
+export const putClientSettings = functions
+  .region("europe-west6")
```
Extract common variable

rjhuijsman replied:
Done.

brunobowden commented on server/functions/src/index.ts:
```diff
@@ -0,0 +1,126 @@
+// Implementation of the v1 API"s `getCaseStats` method.
+// TODO: replace with direct Firestore access from the client.
+export const getCaseStats = functions
+  .region("europe-west6")
```
Extract common variable

rjhuijsman replied:
Done.

brunobowden commented on server/terraform/README.md:
```diff
@@ -167,6 +167,7 @@ which are in the list below. The project renaming is still in progress:
 | who-mh-prod-in-dev | Replicates permissions and setup of who-mh-prod project |
 | who-mh-prod        | WHO Production infrastructure, Privacy Policy compliant |
 | who-mh-staging     | CI server updated from GitHub repo                      |
+| who-mh-v2-dev1     | Development (unstable) environment for v2 backend       |
```
@rjhuijsman 
I was wondering about doing `who-mh2-dev1` instead? I think that'll be less to type and easier to read than an explicit `-v2-`.... which should only be relevant very briefly. You want the developed eye drawn to the `-dev1` which should be the only thing that they need to know.

rjhuijsman replied:
I like that. Done. Note, I still need to delete the `who-mh-v2-dev1` project, I'll leave that until we're 100% sure we don't want it anymore.

brunobowden commented on server/terraform/README.md:
```diff
@@ -182,11 +183,13 @@ emacs main.tf
 terraform init
 terraform apply
-# Should create all resources without any errors
+# THIS MAY THROW AN ERROR:
```
@matthewblain - I'm sure you'll have a comment on this. It is tricky to do this entirely without errors.

brunobowden commented on server/terraform/README.md:
````diff
@@ -257,6 +260,27 @@ Production servers must be configured exactly as described here. Changes should
+
+From the `app/server` folder, run:
+
+```sh
+$ firebase deploy --project=YOUR-PROJECT-ID
+```
````
```suggestion
    firebase deploy --project=YOUR-PROJECT-ID
```

brunobowden commented on server/terraform/modules/myhealth/main.tf:
```diff
@@ -97,11 +98,14 @@ resource "google_firebase_project_location" "fireloc" {
   project       = var.project_id
   location_id   = var.region
-  database_type = "CLOUD_DATASTORE_COMPATIBILITY"
+  database_type = var.backend_version == "v2" ? "CLOUD_FIRESTORE" : "CLOUD_DATASTORE_COMPATIBILITY"
```
No need for backward compatibility, so you can remove the v1 version once v2 is working.... and don't need to wait for the v1 server to be turned down.

rjhuijsman replied:
As per Matthew's suggestion, I've split the v1 and v2 files into separate modules.

brunobowden commented on tools/package-lock.json:
```diff
@@ -1,7 +1,7408 @@
 {
   "name": "app-dev-tools",
+  "lockfileVersion": 2,
```
@theswerd - did we discuss upgrading to lockfileVersion 2 before? Any other implications to worry about?

brunobowden replied:
@theswerd - please comment if there's any issues with lockfileversion 2

rjhuijsman commented on server/.firebaserc:
```diff
@@ -0,0 +1,6 @@
+{
```
I can't leave comments in JSON files, alas, but I've added links to the documentation for the Firebase-specific files (`.firebaserc` and `firebase.json`) to the `README.md`.

rjhuijsman commented on server/.firebaserc:
```diff
@@ -0,0 +1,6 @@
+{
+  "projects": {
+    "default": "who-mh-v2-dev1",
```
When you deploy you'll type `firebase deploy --project=YOUR-PROJECT-ID`, we should always use the `--project=...` flag. The aliases defined here in `.firebaserc` are optional niceties. I've added documentation to the `README.md` about that now.

rjhuijsman commented on server/.gitignore:
```diff
@@ -0,0 +1,62 @@
+# Logs
```
I feel ya.

rjhuijsman commented on server/WHO.postman_collection.json:
```diff
@@ -0,0 +1,174 @@
+{
```
There's a note about this file in the `README.md`, and alas JSON files can't have comments in them themselves. Would you like there to be a more explicit section about "files in this folder" in the `README.md`?

As for using a separate directory: it's not the common practice, but I don't see any strong reasons for or against it either way. Do you have a preference?

rjhuijsman commented on server/fetch-static-serving.sh:
```diff
@@ -0,0 +1,12 @@
+#!/bin/bash
```
Very good questions!
1. I have no idea whether this works on Windows! I'd in fact completely not thought about it! I don't have a Windows machine to test on, alas, and I have 0% knowledge about Windows development - I haven't even used a Windows box since XP... 😮 Do you have the possibility to try it and/or suggest a cross-platform approach?
2. Bazel intuitively does seem like overkill, though it would indeed solve the problem.
3. Localization etc. remains the exact same story as before: the script generates every YAML file in every localization, we continue to upload the same set of ~2000 localized YAML files.

rjhuijsman replied:
I've filed an issue.

rjhuijsman commented on server/functions/package-lock.json:
```diff
@@ -0,0 +1,10617 @@
+{
```
It's more a "make builds reproducible" file than a "review this" file. The tooling updates it automatically.

rjhuijsman commented on server/functions/src/firestore_rules.spec.ts:
```diff
@@ -0,0 +1,41 @@
+import * as firebase from "@firebase/rules-unit-testing";
```
Good question; added comments.

rjhuijsman commented on server/.gitignore:
```diff
@@ -0,0 +1,62 @@
+.grunt
+
+# Bower dependency directory (https://bower.io/)
+bower_components
```
Done.

rjhuijsman commented on server/README.md:
```diff
@@ -19,7 +19,106 @@ Domain: `covid19app.who.int`
+
+Static content is hosted by Firebase Hosting. Everything in the `app/server/public` folder is hosted. Most notable is the `content` folder; the contents of that folder are automatically produced at deploy-time.
+
+Static content will be available via a URL like `https://OUR-PROJECT-ID.web.app` as well as any of our custom domains that we attach to the project (e.g. `sometier.whocoronavirus.org`).
```
Good idea.

rjhuijsman commented on server/README.md:
````diff
@@ -19,7 +19,106 @@ Domain: `covid19app.who.int`
+
+```sh
+app/server$ firebase deploy --project=YOUR-PROJECT-ID
+```
````
Makes sense! Done throughout the file.

brunobowden commented on server/functions/src/index.ts:
```diff
@@ -104,7 +106,7 @@ export const putClientSettings = functions
 // updated settings take effect.
 export const clientSettingsUpdated = functions
-  .region("europe-west6")
+  .region(SERVING_REGION)
```
Is there a way to have a test.... even a deploy time test that barfs if the TypeScript and Terraform regions don't match? They always should but I want to be careful of someone making a stupid mistake.

rjhuijsman replied:
I've added a predeploy test that checks the Cloud Functions region against the region Firestore has been set up for in the project being deployed to.

brunobowden commented on server/README.md:
```diff
@@ -19,7 +19,119 @@ Domain: `covid19app.who.int`
 Domain: `staging.whocoronavirus.org`
 
-## Curl Testing
+### [`who-mh-v2-dev1`(https://console.cloud.google.com/home/dashboard?project=who-mh-v2-dev1)
```
`who-mh2-dev1`, I think you can remove the `mh-v2` reference as you're the only one that's using the project and it will be deleted soon anyway.

brunobowden commented on server/README.md:
````diff
@@ -19,7 +19,119 @@ Domain: `covid19app.who.int`
+```
+
+You should always specify a `--project=YOUR-PROJECT-ID`. If you'd like you can
+use one a [project alias](https://firebase.google.com/docs/cli#project_aliases) instead of a project ID. If you forget to specify a project ID using `--project` Firebase will deploy to the `default` alias, which is aimed at our `who-mh-v2-dev1` project.
````
`who-mh2-dev1`, I think you can remove the `mh-v2` reference as you're the only one that's using the project and it will be deleted soon anyway.

brunobowden commented on server/README.md:
````diff
@@ -19,7 +19,119 @@ Domain: `covid19app.who.int`
+
+To manually experiment with the emulator suite, run the following from the `functions` directory:
+
+```sh
+npm run serve
+```
````
Just indent and don't need the "```sh" wrapper:
```suggestion
    npm run serve
```

brunobowden commented on server/README.md:
```diff
@@ -19,7 +19,119 @@ Domain: `covid19app.who.int`
+
+Currently, the v2 backend isn't ready for use by the client (see [Development Plan](#development-plan)).
+
+### Development Plan
+
+Here are some missing pieces of the v2 backend that you're very welcome to help us implement!
+
+- [ ] Redirect our `index.html` to the app store page based on platform of the requester.
+- [ ] Create a "flavor" in the client that uses our v2 projects. Note the `v1` API is served at a different path within the v2 projects; see `Dynamic content, HTTPS APIs, server-side code` above.
+- [ ] Periodically generate actual case stats (via Cloud Function) and store in Firestore ...
+- [ ] ... And have the client load the actual case stats from Firestore (replace the `getCaseStats` endpoint).
+- [ ] Have the client use the Firestore API directly to register its FCM token (replace the `putClientSettings` endpoint).
+- [ ] Implement the logic that uses updated FCM tokens to register clients for push notifications.
+- [ ] Associate a human-readable domain with our new v2 projects in Firebase Hosting, and implement the appstore redirect.
```
Move to GitHub issues

brunobowden commented on server/terraform/README.md:
```diff
@@ -257,6 +260,27 @@ Production servers must be configured exactly as described here. Changes should
+- Going to the [Firebase Console](console.firebase.google.com)
+- Selecting your newly created project.
+- Clicking the `Cloud Firestore` item in the menu on the left.
+- Clicking the `Create database` button. Leave the settings on default and click `Next`, then `Enable`.
```
@rjhuijsman Seems strange that this can't be automatically done by Terraform?

rjhuijsman replied:
I 100% agree that it's strange! I've spent quite some time trying to make it happen automatically! I wonder if it's a regression. I've reported a bug using Firebase's [form for that](https://firebase.google.com/support/troubleshooter/report/bugs).

brunobowden reviewed (commented):
@rjhuijsman - I'm happy at this point but would want @matthewblain to look it over a again. All my feedback is relatively straight forward.

matthewblain reviewed (commented):
lg

brunobowden commented on server/terraform/modules/myhealth-v2/main.tf:
```diff
@@ -0,0 +1,142 @@
+# MyHealth Service
```
Git just tracks it as a clone. It's only the diff tool that might notice it was a fork.

theswerd commented on server/firestore.rules:
```diff
@@ -0,0 +1,10 @@
+rules_version = '2';
```
Isn't this version #1 for this project?
```suggestion
rules_version = '1';
```

rjhuijsman replied:
Ah, the `rules_version` refers to the version of the _Firebase Rules language_, it's not our own version number.

theswerd commented on server/functions/.eslintrc.js:
```diff
@@ -0,0 +1,27 @@
+  ],
+  plugins: ["@typescript-eslint", "import"],
+  rules: {
+    quotes: ["error", "double"],
```
Love this!

@brunobowden @rjhuijsman thoughts on also banning semicolons?

rjhuijsman replied:
I don't currently have strong feelings either way on semicolons - open to learning why banning would be helpful! 🙂

theswerd commented on server/functions/.prettierrc:
```diff
@@ -0,0 +1,4 @@
+{
+    "trailingComma": "es5",
+    "tabWidth": 2
```
Include single quotes in prettierc
```suggestion
    "tabWidth": 2,
    "singleQuote": true
```

rjhuijsman replied:
So, to clarify: we're currently using double quotes everywhere. The current `eslint` config is set up to follow `prettier`'s default here. You feel we should switch that around?

theswerd commented on server/functions/tsconfig.json:
```diff
@@ -0,0 +1,15 @@
+{
+  "compilerOptions": {
+    "module": "commonjs",
```
Is there a reason to use commonjs over es2020?
```suggestion
    "module": "es2020",
```

rjhuijsman replied:
Google's default Node.js version at the moment is 12, which doesn't do `es2020` modules yet. [Here's a helpful discussion](https://stackoverflow.com/questions/59787574/typescript-tsconfig-settings-for-node-js-12).

theswerd reviewed (changes requested):
Mostly LGTM - Please take a look at comments before merging

rjhuijsman commented on server/firebase.json:
```diff
@@ -0,0 +1,45 @@
+    "hosting": {
+      "port": 5000
+    },
+    "pubsub": {
```
Pub/Sub is not currently used, but will be used in a follow-up PR when we use scheduled Cloud Functions (which are a form of Pub/Sub functions), so it seemed prudent to already introduce the emulator.

Hosting is already used to serve the static content shown in the app.

rjhuijsman commented:
I've addressed all of your comments. I believe we're all happy to merge this version? Would somebody with The Power be willing to press that button @brunobowden @matthewblain @theswerd?

brunobowden commented on server/.firebaserc:
```diff
@@ -0,0 +1,6 @@
+{
+  "projects": {
+    "dev": "who-mh2-dev1",
```
Should we force someone to select dev0, dev1 or others specifically? If we have an alias for dev, then that'll be the one that gets all the collisions.

rjhuijsman replied:
If at some point we introduce a `dev2` project we can remove the alias. Odds are we may only ever have one active `dev` project though.
